File: 2025/regional-andaluza/B/CPP/trie.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

const int ALPHABET_SIZE = 10; // Solo dígitos 0-9

struct TrieNode {
    TrieNode* children[ALPHABET_SIZE];
    int count;

    TrieNode() {
        count = 0;
        for (int i = 0; i < ALPHABET_SIZE; i++)
            children[i] = nullptr;
    }
};

class Trie {
public:
    Trie() {
        root = new TrieNode();
    }

    ~Trie() {
        freeTrie(root);
    }

    void insert(const string& word) {
        TrieNode* node = root;
        for (char ch : word) {
            int index = ch - '0';
            if (node->children[index] == nullptr)
                node->children[index] = new TrieNode();
            node = node->children[index];
            node->count++;
        }
    }

    int countPrefix(const string& prefix) {
        TrieNode* node = root;
        for (char ch : prefix) {
            int index = ch - '0';
            if (node->children[index] == nullptr)
                return 0;
            node = node->children[index];
        }
        return node->count;
    }

private:
    TrieNode* root;

    void freeTrie(TrieNode* node) {
        if (!node) return;
        for (int i = 0; i < ALPHABET_SIZE; i++)
            freeTrie(node->children[i]);
        delete node;
    }
};

int main() {
    int N, Q;
    int caseNumber = 1;

    while (cin >> N >> Q) {
        Trie trie;
        string code;

        for (int i = 0; i < N; i++) {
            cin >> code;
            trie.insert(code);
        }

        cout << "Case " << caseNumber++ << ":\n";
        for (int i = 0; i < Q; i++) {
            cin >> code;
            cout << trie.countPrefix(code) << "\n";
        }
    }

    return 0;
}

File: 2025/regional-andaluza/B/PYTHON/trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, number):
        node = self.root
        for ch in number:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.count += 1

    def count_prefix(self, prefix):
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.count

def main():
    case_number = 1
    try:
        while True:
            line = input()
            if not line:
                continue
            N, Q = map(int, line.strip().split())
            trie = Trie()

            for _ in range(N):
                number = input().strip()
                trie.insert(number)

            print(f"Case {case_number}:")
            for _ in range(Q):
                query = input().strip()
                print(trie.count_prefix(query))

            case_number += 1
    except EOFError:
        pass

if __name__ == "__main__":
    main()
